src/IECoreScene/ShaderNetwork.cpp

```diff
@@ -332,33 +411,62 @@ class ShaderNetwork::Implementation
 			return true;
 		}
 
+
 		void hash( IECore::MurmurHash &h ) const
 		{
-			std::unique_lock<std::mutex> lock( m_hashMutex );
-			if( m_hashDirty )
+			update();
+			h.append( m_hash );
+		}
+
+		void hashSubstitutions( const CompoundObject *attributes, MurmurHash &h ) const
+		{
+			update();
+			for( const auto &a : m_neededSubstitutions )
 			{
-				m_hash = MurmurHash();
-				for( const auto &node : m_nodes )
+				const StringData *sourceAttribute = attributes->member<StringData>( a );
+				if( sourceAttribute )
 				{
-					m_hash.append( node.handle );
-					node.shader->hash( m_hash );
+					h.append( sourceAttribute->readable() );
```

Comment:
Are we sure we want to omit the information about _where_ the substitutions are applied? Might someone reasonably expect the hash to differ if the same substitution was applied to a different parameter? I think this implementation is probably fine (it's what we want in Gaffer I presume), but if so, the current documentation is a bit vague :

```
A hash that tells us whether this network will be affected by any of the given
attributes, based on any requested substitutions
```

Perhaps something more like this?

```
Appends all attributes used by `applySubstitutions()` into the hash.
```